File: 7_kyu/Count_the_Digit.rs
// Count the digit d in the decimal representation of x.
fn count_digit(x: i32, d: i32) -> i32 {
    if x == d {
        // Special case cto overs x=0 and d=d.
        return 1;
    }
    let mut count = 0;
    let mut remain = x.abs();
    while remain > 0 {
        if remain % 10 == d {
            count += 1;
        }
        remain /= 10;
    }
    count
}

fn nb_dig(n: i32, d: i32) -> i32 {
    (0..=n).map(|x| count_digit(x * x, d)).sum()
}

File: 7_kyu/Count_the_Digit.c
#include <stdlib.h>

// Returns count id digit d in the decimal number nr.
static int count_digit(const int nr, const int d)
{
  if( nr == 0 )
    return d == 0 ? 1 : 0;
  int counter = 0;
  int n = abs(nr);
  while( n > 0 )
  {
    if( d == n % 10 )
      ++counter;
    n /= 10;
  }
  return counter;
}

// Returns count of digit d in the nuzmber 0², 1², .. n². 
int nbDig(const int n, const int d)
{
  int counter = 0;
  for( int i = 0; i <= n; ++i )
    counter += count_digit(i*i, d);
  return counter;
}
